### PR TITLE
Fix leaf function patching when requiring red zone

### PR DESCRIPTION
**Pull request by florian-kuebler**
Libunwindstack will remove "speculative" frames, when
two unwinding errors in a row occurs.
There are special situations (ret instruction), where
the unwind information requires reading from the redzone
(rbp in that case), but unwinding actually works perfectly
fine (as rbp is actually not modified). For full unwinding,
libunwindstack will recover and actually completely report
the callstack. However, for leaf function patching, we
only passed in rbp-rsp, so unwinding the second frame will
fail. Libunwindstack will assume the "speculative" frame
was incorrect and removes this.
For the leaf function patching this means, we receive only
one frame from libunwindstack, thus, no need to patch.

This change fixes this issue, by taking a completely new
approach to leaf function patching:
We only unwind one frame from libunwindstack (i.e. the frame
with the instruction pointer), but also read the updated
registers after that unwinding step.
In particular, if rbp was not changed during unwinding,
we know, this was a leaf function, and we can patch in
the updated "pc" entry from the registers, which contains
the "return address".
If, rbp was changed, this could either be an unwinding error,
or a function with frame pointers. We retain our heuristics
to detect unwinding errors and actually send unwinding
errors when non-executable code is on the callstack.

To be still able to handle cases, where the collected
callstack is too small, to actually unwind, but we know
that the sampled function has a frame pointer, we keep
the HasFramePointerSet functionality.

Unfornately, libunwindstack does not report the canonical
frame address. Otherwise, we would have been able to actually
tell if the function had frame pointers set (there could
be cases were CFA does not get computed by $rbp + 16, but
the evaluated value is still $rbp+16).

A potential follow-up would be to improve HasFramePointerSet
such that it computes CFA instead of matching the computation
rule.

Bug: http://b/228428930
Test: Run tests & manually inspect callstacks in samples

### CODE REVIEW

File: src/LinuxTracing/LibunwindstackUnwinder.h [hunk 1]
```diff
@@ -20,24 +20,28 @@
 #include <vector>
 
 #include "OrbitBase/Result.h"
+#include "unwindstack/RegsX86_64.h"
```
Comment (dpallotti) [inline src/LinuxTracing/LibunwindstackUnwinder.h:23]: We include libunwindstack with `<>`.

File: src/LinuxTracing/LeafFunctionCallManager.cpp [hunk 2]
```diff
@@ -69,37 +75,46 @@ Callstack::CallstackType LeafFunctionCallManager::PatchCallerOfLeafFunction(
...
-    ORBIT_ERROR(
-        "Discarding sample as DWARF-based unwinding resulted in empty callstack (error: %s)",
-        LibunwindstackUnwinder::LibunwindstackErrorString(libunwindstack_result.error_code()));
-    return Callstack::kStackTopDwarfUnwindingError;
-  }
+      std::min<uint64_t>(stack_size, stack_dump_size_), true, /*max_frames=*/1);
+
+  // `PatchCallerOfLeafFunction` must not be called for a sample in the outermost frame (_start or
+  // __libc_start_main), so unwinding must not success with max_frames=1.
+  ORBIT_CHECK(!libunwindstack_result.IsSuccess());
```
Comment (dpallotti) [inline src/LinuxTracing/LeafFunctionCallManager.cpp:88]: I'm a bit confused about this check. Where is it guaranteed that we are not calling `PatchCallerOfLeafFunction` for a sample in the outermost frame? And are we sure that libunwindstack couldn't "accidentally" return success in some special case?

File: src/LinuxTracing/LeafFunctionCallManager.cpp [hunk 3]
```diff
@@ -69,37 +75,46 @@ Callstack::CallstackType LeafFunctionCallManager::PatchCallerOfLeafFunction(
...
+
+  unwindstack::RegsX86_64 new_regs = libunwindstack_result.regs();
 
-  if (libunwindstack_callstack.size() == 1) {
-    if (stack_size > stack_dump_size_ && !libunwindstack_result.IsSuccess()) {
+  // If both pc, and $rsp does not change during unwinding, there was an unwinding error.
```
Comment (dpallotti) [inline src/LinuxTracing/LeafFunctionCallManager.cpp:92]:
```suggestion
  // If both pc and $rsp do not change during unwinding, there was an unwinding error.
```

File: src/LinuxTracing/LeafFunctionCallManager.cpp [hunk 4]
```diff
@@ -69,37 +75,46 @@ Callstack::CallstackType LeafFunctionCallManager::PatchCallerOfLeafFunction(
...
-    return Callstack::kComplete;
+    return Callstack::kStackTopDwarfUnwindingError;
   }
 
-  // If unwinding results in more than two frames, $rbp was also not set correctly by the caller,
-  // thus frame pointers are not all in non-leaf functions, and our assumptions do not hold.
-  if (libunwindstack_callstack.size() > 2) {
-    return Callstack::kFramePointerUnwindingError;
+  uint64_t new_rbp = new_regs[unwindstack::X86_64_REG_RBP];
+  // $rbp changed during unwinding (case (2)), i.e. it was a valid frame pointer and thus the
```
Comment (dpallotti) [inline src/LinuxTracing/LeafFunctionCallManager.cpp:103]:
```suggestion
  // $rbp changed during unwinding (case (2)), i.e. either it was a valid frame pointer and thus the
```

File: src/LinuxTracing/LeafFunctionCallManager.cpp [hunk 5]
```diff
@@ -69,37 +75,46 @@ Callstack::CallstackType LeafFunctionCallManager::PatchCallerOfLeafFunction(
...
+    // $rbp could not be a valid frame pointer.
+    // Note, that in addition to this check, we also check if the complete callchain is in
+    // executable code in the UprobesUnwindingVisitor.
+    if (new_rbp < rbp) {
+      return Callstack::kFramePointerUnwindingError;
+    }
+    return Callstack::kComplete;
   }
 
+  // $rbp did non changed during unwinding, i.e. we are in a leaf function. We need to patch in the
```
Comment (dpallotti) [inline src/LinuxTracing/LeafFunctionCallManager.cpp:116]:
```suggestion
  // $rbp did non change during unwinding, i.e. we are in a leaf function. We need to patch in the
```

File: src/LinuxTracing/LeafFunctionCallManager.cpp (hunk 4 above)
Comment (florian-kuebler) [reply src/LinuxTracing/LeafFunctionCallManager.cpp]: Done.

File: src/LinuxTracing/LeafFunctionCallManager.cpp [hunk 6]
```diff
@@ -69,37 +75,46 @@ Callstack::CallstackType LeafFunctionCallManager::PatchCallerOfLeafFunction(
...
 
-  // If unwinding results in more than two frames, $rbp was also not set correctly by the caller,
-  // thus frame pointers are not all in non-leaf functions, and our assumptions do not hold.
-  if (libunwindstack_callstack.size() > 2) {
-    return Callstack::kFramePointerUnwindingError;
+  uint64_t new_rbp = new_regs[unwindstack::X86_64_REG_RBP];
+  // $rbp changed during unwinding (case (2)), i.e. it was a valid frame pointer and thus the
+  // callchain is already correct, or it was modified as general purpose register (unwinding error).
+  if (new_rbp != rbp) {
+    // If the $rbp after unwinding is below (downwards growing stack) the sampled $rbp, the sampled
```
[the remark this answers is not in the document]
Comment (florian-kuebler) [reply src/LinuxTracing/LeafFunctionCallManager.cpp]: Moved the comment.

File: src/LinuxTracing/LeafFunctionCallManager.cpp (hunk 5 above)
Comment (florian-kuebler) [reply src/LinuxTracing/LeafFunctionCallManager.cpp]: Done.

File: src/LinuxTracing/LeafFunctionCallManager.cpp [hunk 7]
```diff
@@ -24,16 +24,22 @@ namespace orbit_linux_tracing {
...
+// updated and we can detect if $rbp was modified.
+// (1) If $rbp did not change: We are in a leaf function, which has not modified $rbp. The leaf's
+//     caller is missing in the callchain and needs to be patched in. The updated $rip (pc) from
+//     the unwinding step contains the leaf's caller.
+// (2) If $rbp was modified, this can either be:
+//     (a) We are in a non-leaf function and the callchain is already correct.
+//     (b) We are in a leaf function that modified $rbp. The complete callchain is broken and should
+//         be reported as unwinding error.
+// As libunwindstack does not report us the canonical frame address (CFA) from an unwinding step, we
+// can not differentiate between (2a) and (2b) reliably. However, we do perform the following
```
Comment (dpallotti) [inline src/LinuxTracing/LeafFunctionCallManager.cpp:37]:
```suggestion
// cannot differentiate between (2a) and (2b) reliably. However, we do perform the following
```

File: src/LinuxTracing/LeafFunctionCallManager.cpp [hunk 8]
```diff
@@ -69,58 +74,66 @@ Callstack::CallstackType LeafFunctionCallManager::PatchCallerOfLeafFunction(
...
-      libunwindstack_result.frames();
-
-  if (libunwindstack_callstack.empty()) {
-    ORBIT_ERROR(
-        "Discarding sample as DWARF-based unwinding resulted in empty callstack (error: %s)",
-        LibunwindstackUnwinder::LibunwindstackErrorString(libunwindstack_result.error_code()));
-    return Callstack::kStackTopDwarfUnwindingError;
+      std::min<uint64_t>(stack_size, stack_dump_size_), true, /*max_frames=*/1);
+
+  // If unwinding a single frame yields a success, we are in the outer-most frame, i.e. don't have
```
Comment (dpallotti) [inline src/LinuxTracing/LeafFunctionCallManager.cpp:85]:
```suggestion
  // If unwinding a single frame yields a success, we are in the outer-most frame, i.e. we don't have
```

File: src/LinuxTracing/LeafFunctionCallManager.cpp [hunk 9]
```diff
@@ -69,58 +74,66 @@ Callstack::CallstackType LeafFunctionCallManager::PatchCallerOfLeafFunction(
...
-  // thus frame pointers are not all in non-leaf functions, and our assumptions do not hold.
-  if (libunwindstack_callstack.size() > 2) {
-    return Callstack::kFramePointerUnwindingError;
+  uint64_t new_rbp = new_regs[unwindstack::X86_64_REG_RBP];
+  // $rbp changed during unwinding (case (2)), i.e. either it was a valid frame pointer and thus the
+  // callchain is already correct, or it was modified as general purpose register (unwinding error).
+  if (new_rbp != rbp) {
+    // If the $rbp after unwinding is below the sampled $rbp, the sampled $rbp could not be a valid
+    // frame pointer (remember the stack grows downwards).
+    // Note, that in addition to this check, we also check if the complete callchain is in
```
Comment (dpallotti) [inline src/LinuxTracing/LeafFunctionCallManager.cpp:109]:
```suggestion
    // Note that, in addition to this check, we also check if the complete callchain is in
```

**Review by dpallotti — APPROVED** [review]
Great, just me picking on the comments.

File: src/LinuxTracing/LeafFunctionCallManager.cpp (hunk 7 above)
Comment (florian-kuebler) [reply src/LinuxTracing/LeafFunctionCallManager.cpp]: Done.

File: src/LinuxTracing/LeafFunctionCallManager.cpp (hunk 8 above)
Comment (florian-kuebler) [reply src/LinuxTracing/LeafFunctionCallManager.cpp]: Done.

File: src/LinuxTracing/LeafFunctionCallManager.cpp (hunk 9 above)
Comment (florian-kuebler) [reply src/LinuxTracing/LeafFunctionCallManager.cpp]: Done.

**Review by florian-kuebler — COMMENTED** [review]
Thanks

**Pull request merged**